packages/ingest/src/HttpProviderClient.js

```diff
@@ -62,14 +64,35 @@ class HttpProviderClient {
 
     if (this.username) this.gotOptions.username = this.username;
     if (this.password) this.gotOptions.password = this.password;
+
+    const RedirectHandler = {
+      // Need to use named function and not fat arrow
+      // expression so that we can use the bound value
+      // of "this"
+      handleBeforeRedirect(options, response) {
+        if (redirectCodes.has(response.statusCode)) {
```

Comment:
The question is: is this secure? `got` is intentionally disabling the forwarding of credentials when redirecting to a different host, which is good for security. Otherwise you are forwarding your basic auth credentials to a site that may or may not need them and could in theory be a bad actor that intercepts them.

The only way to keep the forwarding of credentials on redirect is maybe to add a new provider type, `basic-auth`, and another provider field of `redirectAuthHost`, so then you can explicitly set what host(s) should be used to allow auth redirects and avoid sending credentials to possibly malicious sites. Thoughts?

Reply:
So you have a request coming from sync-granule which includes a Basic Auth header because it's a workflow task and we can't prompt for login or anything like that. It redirects to EDL but the basic auth header has been dropped, then EDL responds with a 401 or it redirects back to the distribution API which responds with a 401 (doesn't matter I guess). Is that basically the problem?

A whitelist with our auth providers (currently EDL and Cognito) is one way to go but that does seem like something that could be forgotten about if another auth provider is added. I'm not sure how easily catchable that would be but that might be the best solution. I can't imagine adding auth providers is something that's going to happen regularly and maybe it would be possible to write a test that would fail if you add a new one and don't whitelist it here.

Is there any way to avoid the redirect altogether or is that a major change (seems like yes)? If we separated the steps where one request makes the auth call and another subsequent request either includes the cookie or uses the existing session to download the files?

Reply:
>Is that basically the problem?

Yes, that's the problem

>A whitelist with our auth providers (currently EDL and Cognito) is one way to go but that does seem like something that could be forgotten about if another auth provider is added

I'm a little worried about whitelisting only specific auth providers because it just seems likely that either one of the DAACs or our non-NASA users will need to download files from a provider protected by some other auth service. So I'm guess I'm leaning towards making it explicit as a field on the provider.

> Is there any way to avoid the redirect altogether or is that a major change (seems like yes)? If we separated the steps where one request makes the auth call and another subsequent request either includes the cookie or uses the existing session to download the files?

I don't see how we can avoid the redirect whether we separate the request into two steps or not. The HTTP provider itself is what dictates the file being protected by an auth service and so we can't avoid it forcing a redirect on download

You make a good point that I should ensure that a cookie session is set/respected via unit testing, but I believe it will be. However, for the initial request before a session cookie is established, I'm not sure I understand the benefit of separating into two requests vs one

Reply:
I decided to implement a `basicAuthRedirectHost` field on provider and only forward credentials on redirect if the host for the redirect matches `basicAuthRedirectHost`

Reply:
The technical approach seems sound.   The security concern here, in theory, is exposure of EDL credentials via a compromised provider host redirecting to <somewhere else>.... this seems to cover that well enough. 

A couple of thoughts: 

---
> I'm a little worried about whitelisting only specific auth providers because it just seems likely that either one of the DAACs or our non-NASA users will need to download files from a provider protected by some other auth service. So I'm guess I'm leaning towards making it explicit as a field on the provider.

This could avoid adding another key to the provider object if it's a global configuration setting *or* a workflow configuration option for task components that might require it.      Something to consider. 
---

- More generally -- is there a reason not to explicitly fail the request on an unexpected redirect?      Perhaps more to the point, is there generally a good reason other than auth that a http provider *should* redirect other than internally?

Reply:
>This could avoid adding another key to the provider object if it's a global configuration setting or a workflow configuration option for task components that might require it. Something to consider.

It seems like making it a workflow config is less flexible though - because then it is the same allowed redirect for any invocation of that workflow, whereas a provider field means it can be different for any provider used with that workflow. Absent another benefit to using the workflow config approach, I'm inclined to stick with adding the provider property.

> More generally -- is there a reason not to explicitly fail the request on an unexpected redirect? Perhaps more to the point, is there generally a good reason other than auth that a http provider should redirect other than to an alternate location on the same host?

Can we safely make those kind of assumptions? I mean what if a DAAC for some reason is moving their holdings to a new domain and putting in 301 redirects to get people to the new data location?

Reply:
> Can we safely make those kind of assumptions? I mean what if a DAAC for some reason is moving their holdings to a new domain and putting in 301 redirects to get people to the new data location?

I suppose not given the current implementation, although if we make the 'allowed redirect' a *list* that would work, even if it's probably one of my making in the above comment.  

So, yes, we could write that such that it could be adaptable to that situation.     Is it needed/value added?      Up for debate.     We might want to consider a list regardless of our current implementation though.

Reply:
So we could just generally have an `allowedRedirects` property and deny any redirects outside of that? And forward credentials to any hosts in that list?

Reply:
Yes, that's what I was suggesting.    The only hiccup there is if redirects are prevalent in the wild and/or having to explicitly define them places an absurd burden on the end user. 

If you don't hate this idea we should probably whiteboard it briefly

If for some reason we decide we don't like rejecting redirects wholesale at a minimum I think we should consider `allowedRedirects` as a list instead of a single string for reasons of flexibility in the data model/situations we haven't anticipated.

Reply:
We talked over the approach in the 21 Jul Arch meeting with the assembled team.  The updates made were considered an improvement on the existing (failed) behavior.